text/php/0002-buildpack-yml-to-env-var.md

```diff
@@ -0,0 +1,222 @@
+```shell
+$BP_PHP_WEB_DIR="htdocs"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+php:
+ webdirectory: htdocs
+```
+
+### `BP_PHP_LIB_DIR`
+```shell
+$BP_PHP_LIB_DIR="lib"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+php:
+ libdirectory: lib
+```
+
+### `BP_PHP_SERVER_ADMIN`
+```shell
+$BP_PHP_SERVER_ADMIN="admin@localhost"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+php:
+ serveradmin: admin@localhost
+```
+
+### `BP_REDIS_SESSION_STORE_SERVICE_NAME`
+```shell
+$BP_REDIS_SESSION_STORE_SERVICE_NAME="redis-session"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+php:
+ redis:
+   session_store_service_name: redis-sessions
+```
+
+### `BP_MEMCACHED_SESSION_STORE_SERVICE_NAME`
+```shell
+$BP_MEMCACHED_SESSION_STORE_SERVICE_NAME="memcached-session"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+php:
+ memcached:
+   session_store_service_name: memcached-sessions
+```
+
+### Configuration Removal
+The following structure in `buildpack.yml` will not be receiving an environment
+variable configuration option:
+```yaml
+php:
+ script:
+```
+This structure was effectively taking the part and behavior of [Paketo Procfile
+buildpack](https://github.com/paketo-buildpacks/procfile) by allowing users to
+provide a start command in `buildpack.yml`. However, as part of the [PHP Rewrite RFC](https://github.com/paketo-buildpacks/rfcs/blob/main/text/php/0001-restructure.md)
+the Procfile buildpack will be added to all order groupings. Because of this
+addition I am proposing the removal of this configuration option in the
+environment variables in favor of encouraging users to use a Procfile if they
+need to set a custom start script.
+
+---
+The proposed environment variables for Composer are as follows:
+
+### `BP_COMPOSER_VERSION`
+```shell
+$BP_COMPOSER_VERSION="1.10.x"
+```
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+composer:
+  version: 1.10.x
+```
+### `BP_COMPOSER_INSTALL_OPTIONS`
+```shell
+$BP_COMPOSER_INSTALL_OPTIONS="--no-dev --prefer-install=auto"
+```
+
+Note: This will be parsed using this [shellwords library](https://github.com/mattn/go-shellwords).
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+composer:
+  install_options: ["--no-dev", "--prefer-install=auto"]
+```
+### `BP_COMPOSER_VENDOR_DIR`
+```shell
+$BP_COMPOSER_VENDOR_DIR="vendor"
+```
+
+Note: The value of this environment variable will be used to set
+[`$COMPOSER_VENDOR_DIR`](https://getcomposer.org/doc/03-cli.md#composer-vendor-dir)
+therefore if a user attempts to set this value themselves as well as set the
+value of `$BP_COMPOSER_VENDOR_DIR` the value of `$COMPOSER_VENDOR_DIR` will be
+overwritten resulting in potentially unexpected behavior.
+
+This will replace the following structure in `buildpack.yml`:
+```yaml
+composer:
+  vendor_directory: vendor
+```
```

Comment:
This seems unnecessary. As the comment mentions, it could result in confusion too. Which one does a user set `COMPOSER_VENDOR_DIR` or `BP_COMPOSER_VENDOR_DIR`. It seems like we could just drop the `BP_` version and if a user wants to set the vendor dir, then they set that specifically. 

I generally like to use existing env variable options when available and since this is just being passed through to composer, that seems reasonable here.

Thoughts?

Reply:
I think that the only potential issue is that currently we set these relative the working directory by joining the working directory with the given value. We would lose that functionality of prepending the working directory and I do not know if composer would be able to handle that.

Reply:
Yes, that's a concern. The other issue here might be with caching. The actual vendor directory should probably be to a layer so we can cache it, while Composer and the app need to see a symlink in the `/workspace` because that is what they think.

I still think we should probably use the default env variable, but we may need to do some special preprocessing on it before we pass it through to Composer. Probably use that as the location for the symlink, since from the end-user perspective that's what they care about.

Reply:
I have some concerns using a native configuration option and then doing preprocessing in the buildpack. I think that doing this leads to confusion and unexpected behavior from users that are expecting things to work identically. I think that there are things that we could do to make the processing very transparent to the end user I just want your take on the drawbacks in my opinion before going in and changing it.

Reply:
I think as long as we can guarantee two things it's ok.  1.) whatever additional processing we're doing is transparent to the end user and 2.) the setting works just like the user would typically expect when running locally or on a VM.

My take is that it should. If a user sets `COMPOSER_VENDOR_DIR` locally, that changes the directory where composer will write those files. In the buildpack, I think what we'll need to do is read `COMPOSER_VENDOR_DIR` and add a symlink to our build+cache layer at the `COMPOSER_VENDOR_DIR` location. That way when composer runs, it follows the symlink to the layer but at the same time things just look like the vendor dir already exists. Composer should add stuff to the specified location and it'll "just work". That's my theory on it at least.

The whole thing is tricky though because many PHP apps expect the vendor directory to exist along side the application files, which is in `/workspace`. We can't write them directly there, otherwise we need to run `composer install` every time which is slow or copy files from the cache layer to `/workspace` which is also slow. In this case, we just need to recreate the symlink every time, which is fast.

Reply:
Alright I have updated this segment to hopefully include everything that would be needed in this RFC. Please take one more look over when you have time.